flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/observer/deployment/AbstractDeploymentObserver.java

```diff
@@ -279,21 +279,35 @@ private void checkIfAlreadyUpgraded(FlinkDeployment flinkDep, Context context) {
                     Long upgradeTargetGeneration =
                             ReconciliationUtils.getUpgradeTargetGeneration(flinkDep);
 
+                    Long currentSpecGeneration = flinkDep.getMetadata().getGeneration();
+
                     if (deployedGeneration.equals(upgradeTargetGeneration)) {
-                        logger.info(
-                                "Last reconciled generation is already deployed, setting reconciliation status to "
-                                        + ReconciliationState.DEPLOYED);
+                        logger.info("Pending upgrade is already deployed, updating status.");
 
                         var firstDeploy =
                                 status.getReconciliationStatus().getLastReconciledSpec() == null;
-                        var conf =
-                                firstDeploy
-                                        ? configManager.getDeployConfig(
-                                                flinkDep.getMetadata(), flinkDep.getSpec())
-                                        : configManager.getObserveConfig(flinkDep);
+                        if (firstDeploy) {
+                            ReconciliationUtils.updateForSpecReconciliationSuccess(
+                                    flinkDep,
+                                    JobState.RUNNING,
+                                    configManager.getDeployConfig(
+                                            flinkDep.getMetadata(), flinkDep.getSpec()));
+                        } else {
+                            ReconciliationUtils.updateStatusForAlreadyUpgraded(flinkDep);
+                        }
+
+                        status.getJobStatus()
+                                .setState(org.apache.flink.api.common.JobStatus.RECONCILING.name());
+                        status.setJobManagerDeploymentStatus(JobManagerDeploymentStatus.DEPLOYING);
+                    } else if (deployedGeneration.equals(currentSpecGeneration)) {
```

Comment:
I think we have a problem here, we can only do the thing when the deployedGeneration equal to the `currentSpecGeneration` or the `upgradeTargetGeneration`. If they are not equal, then we do not know how to bring the job do the desired version.

Reply:
What about also record the updateTargetSpec before execute the upgrade or submit. So that, we can always have the right base to compare. And it will not changed by user. In the observer, we compare it with the deployedGeneration. WDYT?

Reply:
this only affects the first deployment scenario so it’s not a critical optimization but I can see how complicated it would be to add

Reply:
I will check this today